### DISCUSSION TITLE
Guidance on Git, branching strategy and releases

### DISCUSSION THREAD

**Discussion by nhsbsa-pattu** [nhsbsa/nhsbsa-digital-playbook#375] [Development]
We need guidance on our preferred approach to Git branching, and subsequent release.

We should cover:

* Git flows we align to (we don't use gitlab flow)
* Code review
  We must have code review prior to release
  Link to code review process
* Protected branches
* `main` branch must reflect Production
  Must have `ff-only` when merging into `main`
* Hotfixing
* Tag to release
  Discuss auto-bumping?
* Release condidates
  Deploy to test env
* Naming conventions

# References

* https://bsa2468.atlassian.net/wiki/spaces/HRT/pages/3979313153/Git+Commit+merging+and+branching+strategy
* https://bsa2468.atlassian.net/wiki/spaces/PROJBSA/pages/4496949252/Branching+strategy+-+BSA+Cloud
* https://bsa2468.atlassian.net/wiki/spaces/NJA/pages/4567728141/NHS+Jobs+V2+Release+Branching+Strategy

**Comment by LouisRoberts** [comment]
I don't have a work GitHub account, so using my personal account to comment.
A few things to think about -
**Config**. It is not 'just' the repository, there is also config \ terraform changes that are part of a release \ deployment, but outside the repository. We have had a lot of changes to terraform 'or' the base terraform scripts that requires a redeployment - but the main code base has stayed the same. 
Config needs to be aligned somehow. If we merge in changes to deploy, we should be able to revert i.e. redeploy with the previous config.
**Multiple repositories forming a deployment**. I think somebody mentioned that you can create a '**milestone**' in GitLab, so maybe need to look at that as a way of grouping a set of repositories together. Some changes apply to a 'group' of repositories. Managing this is complicated. You also have to consider when deploying what happens if a single component fails and you have to rollback the lot. Blue\Green deployment is great for a single repository, but when you have to deploy a set of micro services for a single feature, then we need to think how the build pipeline works in this instance. Database changes in micro services are also complex when we have to revert. We shouldn't be bringing the whole service down due to the complexity of the deployment when more than 1 repository is deployed. 
**Fast Forward Strategy**. We need guidance on how to fast forward (guessing this is what ff above is) works. So far every time I hit this, I am unable to merge the changes in and have to create 'another' branch from the branch I am trying to merge into and reapply my changes, losing the history. Trying to do this locally on the command line is very difficult. We need a better tool to manage such merges. Oddly it works better when we have merge conflicts as we can pick the changes online than when fast forward fails.
**Pre conditions for pushing changes** The rules are useful to make you think about the checks when pushing code. The issue is when a push fails from an IDE to a pre condition failing. There is little help on why it has failed. You also get issues when trying to merge after a review as the pre condition prevents the merge, but doesn't tell you why. Quite often it is a syntactic issue or a name issue. However once you have a commit message that doesn't follow 'the rules' , you have to abandon that branch and start over.

Remember Agile is about 'People over Process'. The current process has some nice features, but has some ruff edges. So part of the guidance should include making the current process better.

**Reply by nhsbsa-pattu** [reply]
Thanks Louis. Feel free to create a NHSBSA account in Github, although its not mandated.

I think we should separate our branching and versioning strategy from our deployment strategy. I.e. we can branch for the next changes and then release that component under the next semver. Deployment into production requires coordination as you've said, but the scope of a branch  is only within the repo. So the coordinated deployment needs to be managed elsewhere.

NHS Jobs has approached this problem using a 'release' repository. It contains a manifest file that declares explicitly which version of what component is being released into which environment. For candidate releases, it declares a branch name, rather than a semver. It works pretty well. 

I have started a new discussion on CI/CD to cover the deployment aspect, along with other aspects. We may need a few pages to cover this. #379

On Fast Forward, it sounds like some Git training is needed. I think we can run some internal workshops, maybe some drop-in sessions?
